x/campaign/keeper/msg_server_add_mainnet_vesting_account.go

```diff
@@ -0,0 +1,53 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/tendermint/spn/x/campaign/types"
+	profiletypes "github.com/tendermint/spn/x/profile/types"
+)
+
+func (k msgServer) AddMainnetVestingAccount(goCtx context.Context, msg *types.MsgAddMainnetVestingAccount) (*types.MsgAddMainnetVestingAccountResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	campaign, found := k.GetCampaign(ctx, msg.CampaignID)
+	if !found {
+		return nil, sdkerrors.Wrapf(types.ErrCampaignNotFound, "%v", msg.CampaignID)
+	}
+
+	// Get the coordinator ID
+	coordinatorID, found := k.profileKeeper.CoordinatorIDFromAddress(ctx, msg.Coordinator)
+	if !found {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordAddressNotFound, msg.Coordinator)
+	}
+	if campaign.CoordinatorID != coordinatorID {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordInvalid, fmt.Sprintf(
+			"coordinator of the campaign is %v",
+			campaign.CoordinatorID,
+		))
+	}
+
+	if campaign.MainnetInitialized {
+		return nil, sdkerrors.Wrapf(types.ErrMainnetInitialized, "%v", msg.CampaignID)
+	}
+
+	account := types.MainnetVestingAccount{
+		CampaignID:     campaign.Id,
+		Address:        msg.Address,
+		Shares:         msg.Shares,
+		VestingOptions: msg.VestingOptions,
+	}
+
+	campaign.AllocatedShares = types.IncreaseShares(campaign.AllocatedShares, msg.Shares)
```

Comment:
The shares in `VestingOptions` must also be taken into account for the allocation share increase of the campaign